### PR TITLE
Broker working with controller and monitor. Interactive client and ro…

### PR DESCRIPTION
**Pull request by betobetodev**
…utes to logs added

### CODE REVIEW

File: application_manager/service/api/v10.py [hunk 1]
```diff
@@ -57,12 +57,34 @@ def status():
...
+
+    stderr.close()
+    stdout.close()
+
+    return out, err
+
 def _get_new_cluster_size(hosts):
     return optimizer.get_cluster_size(api.optimizer_url, hosts)
 
+def _replace(string):
```
Comment (armstrongmsg) [inline application_manager/service/api/v10.py]: maybe rename to "_remove_newline"?

File: application_manager/plugins/sahara/plugin.py [hunk 2]
```diff
@@ -236,7 +240,7 @@ def _wait_on_job_finish(self, sahara, connector, job_exec_id,
             current_time = datetime.datetime.now()
             current_job_time = (current_time - start_time).total_seconds()
             if current_job_time > 3600:
```
Comment (armstrongmsg) [inline application_manager/plugins/sahara/plugin.py:258]: This if block should be removed. It does not make sense to kill an application that takes more than one hour to finish.

File: application_manager/service/api/v10.py [hunk 3]
```diff
@@ -57,12 +57,34 @@ def status():
         application_stat["status"] = applications[app_id].get_application_state()
         application_stat["time"] = applications[app_id].get_application_execution_time()
         application_stat["start_time"] = applications[app_id].get_application_start_time()
-
+    
     return applications_status
 
+def broker_log():
+    log = open("logs/sahara_plugin.log", "r")
```
Comment (armstrongmsg) [inline application_manager/service/api/v10.py]: I believe this log file is only used in Spark Sahara plugin. How about OpenStack Generic?

File: application_manager/service/api/v10.py [hunk 4]
```diff
@@ -57,12 +57,34 @@ def status():
...
+    str_log = map(_replace, log.readlines())
```
Comment (armstrongmsg) [inline application_manager/service/api/v10.py]: What if the log file is too big?

File: application_manager/plugins/sahara/plugin.py [hunk 5]
```diff
@@ -69,6 +69,10 @@ def get_application_start_time(self):
 
     def start_application(self, data, spark_applications_ids, app_id):
         try:
+            # Clean sahara log file
+            f = open("logs/sahara_plugin.log", "w")
```
Comment (armstrongmsg) [inline application_manager/plugins/sahara/plugin.py]: What if many applications are running at the same time? Maybe we should use different log files for each application. Something like "sahara_plugin_{app_id}.log".

**Review by armstrongmsg — CHANGES_REQUESTED** [review]

File: application_manager/plugins/sahara/plugin.py [hunk 6]
```diff
@@ -463,13 +463,11 @@ def _submit_job(self, remote_instance, key_path, main_class,
...
             spark_submit = spark_submit.replace('--packages', '')
-
         job = remote.execute_command_popen(remote_instance, key_path, 
                                            spark_submit)
 
         return job
-
-        
+      
     def _mkdir(self, path):
```
Comment (armstrongmsg) [inline application_manager/plugins/sahara/plugin.py]: Why not use this instead? https://docs.python.org/2/library/os.html#os.mkdir

**Review by armstrongmsg — CHANGES_REQUESTED** [review]

File: application_manager/plugins/sahara/plugin.py (hunk 2 above)
Comment (tellesnobrega) [reply application_manager/plugins/sahara/plugin.py:258]: I think we could add a timeout but it should be bigger.

File: application_manager/service/api/v10.py (hunk 1 above)
Comment (tellesnobrega) [reply application_manager/service/api/v10.py]: agreed

File: application_manager/plugins/sahara/plugin.py [hunk 7]
```diff
@@ -378,26 +383,58 @@ def _hdfs_spark_execution(self, master, remote_hdfs, key_path, args,
         self._mkdir(local_path)
 
         # Create cluster directories
-        plugin_log.log("%s | Creating cluster directories" %
+        plugin_log.log("%s | Creating directories (CLUSTER)" %
```
Comment (tellesnobrega) [inline application_manager/plugins/sahara/plugin.py]: porque essa mudança?

File: application_manager/plugins/sahara/plugin.py (hunk 2 above)
Comment (betobetodev) [reply application_manager/plugins/sahara/plugin.py:258]: Should this be parametrized by the client?

Comment (armstrongmsg) [reply application_manager/plugins/sahara/plugin.py:258]: Yes, I think it is a good solution to add the timeout as parameter.

File: application_manager/utils/remote.py [hunk 8]
```diff
@@ -43,11 +43,19 @@ def read_file_from(remote_file, run_as_root=False, timeout=120):
...
-    command = ('ssh -i %(key_path)s ubuntu@%(remote)s %(command)s' %
+    command = ('ssh -o "StrictHostKeyChecking no" -o "UserKnownHostsFile=/dev/null" -i %(key_path)s ubuntu@%(remote)s %(command)s' %
                {'remote': remote, 'key_path': key_path, 'command': command})
 
     subprocess.call(command, shell=True)
 
+def execute_command_popen(remote, key_path, command):
+    command = ('ssh -o "StrictHostKeyChecking no" -o "UserKnownHostsFile=/dev/null" -i %(key_path)s ubuntu@%(remote)s %(command)s' %
+               {'remote': remote, 'key_path': key_path, 'command': command})
+
```
Comment (armstrongmsg) [inline application_manager/utils/remote.py]: Remove this blank line

File: application_manager/utils/remote.py [hunk 9]
```diff
@@ -43,11 +43,19 @@ def read_file_from(remote_file, run_as_root=False, timeout=120):
 
 
 def execute_command(remote, key_path, command):
-    command = ('ssh -i %(key_path)s ubuntu@%(remote)s %(command)s' %
+    command = ('ssh -o "StrictHostKeyChecking no" -o "UserKnownHostsFile=/dev/null" -i %(key_path)s ubuntu@%(remote)s %(command)s' %
```
Comment (armstrongmsg) [inline application_manager/utils/remote.py:46]: Isn't "StrictHostKeyChecking no" enough? Why do you need to pass /dev/null as known_hosts file?

File: application_manager/api/v10.py [hunk 10]
```diff
@@ -23,18 +23,24 @@
...
+def status():
     return u.render(api.status())
+
+@rest.get('/manager/logs/execution/<app_id>')
+def execution_log(app_id):
+    return u.render(api.execution_log(app_id))
+
+@rest.get('/manager/logs/std/<app_id>')
+def std_log(app_id):
+    return u.render(api.std_log(app_id))
```
Comment (armstrongmsg) [inline application_manager/api/v10.py:46]: How about the standard error output? Why isn't there a "stderr_log" option?

**Review by armstrongmsg — CHANGES_REQUESTED** [review]

File: application_manager/api/v10.py (hunk 10 above)
Comment (betobetodev) [reply application_manager/api/v10.py:46]: 'std_log' provides the standard outputs of a Spark submission: stderr and stdout.

File: application_manager/utils/remote.py (hunk 9 above)
Comment (betobetodev) [reply application_manager/utils/remote.py:46]: 'UserKnownHostsFile=/dev/null' ignores the "man-in-the-middle" message.

Comment (armstrongmsg) [reply application_manager/utils/remote.py:46]: ok

File: application_manager/api/v10.py (hunk 10 above)
Comment (iurygregory) [reply application_manager/api/v10.py:46]: std_log it's fine since is for both

File: application_manager/utils/remote.py (hunk 8 above)
Comment (betobetodev) [reply application_manager/utils/remote.py]: Removed.

**Pull request merged**